src/ray/raylet/task_dependency_manager.cc

```diff
@@ -180,33 +229,50 @@ bool TaskDependencyManager::UnsubscribeDependencies(const TaskID &task_id) {
   task_dependencies_.erase(it);
 
   // Remove the task's dependencies.
-  for (const auto &object_id : task_entry.object_dependencies) {
-    // Remove the task from the list of tasks that are dependent on this
-    // object.
-    // Get the ID of the task that creates the dependency.
-    TaskID creating_task_id = ComputeTaskId(object_id);
-    auto creating_task_entry = required_tasks_.find(creating_task_id);
-    std::vector<TaskID> &dependent_tasks = creating_task_entry->second[object_id];
-    auto it = std::find(dependent_tasks.begin(), dependent_tasks.end(), task_id);
-    RAY_CHECK(it != dependent_tasks.end());
-    dependent_tasks.erase(it);
-    // If the unsubscribed task was the only task dependent on the object, then
-    // erase the object entry.
-    if (dependent_tasks.empty()) {
-      creating_task_entry->second.erase(object_id);
-      // Remove the task that creates this object if there are no more object
-      // dependencies created by the task.
-      if (creating_task_entry->second.empty()) {
-        required_tasks_.erase(creating_task_entry);
-      }
-    }
+  for (const auto &object_id : task_entry.get_dependencies) {
+    RemoveTaskDependency(task_id, object_id);
+  }
+  for (const auto &object_id : task_entry.wait_dependencies) {
+    RemoveTaskDependency(task_id, object_id);
   }
 
   // These dependencies are no longer required by the given task. Cancel any
   // in-progress operations to make them local.
-  for (const auto &object_id : task_entry.object_dependencies) {
+  for (const auto &object_id : task_entry.get_dependencies) {
     HandleRemoteDependencyCanceled(object_id);
   }
+  for (const auto &object_id : task_entry.wait_dependencies) {
+    HandleRemoteDependencyCanceled(object_id);
+  }
+
+  return true;
+}
+
+bool TaskDependencyManager::UnsubscribeDependency(const TaskID &task_id, const ObjectID &object_id) {
```

Comment:
I would change this method so that it only removes the `object_id` from `wait_dependencies`, since we already have a method for unsubscribing from `get_dependencies`. Also, we should make this method private and call it inside `TaskDependencyManager::HandleObjectLocal`, to clean up `ray.wait` calls that get fulfilled after the initial call has already returned.